File: algorithms/lab08/python/01-dijkstra/01-dijkstra.py
# SPDX-License-Identifier: BSD-3-Clause

import heapq


class Task:
    def __init__(self) -> None:
        # numarul maxim de noduri
        self.NMAX = 50005

        # valoare mai mare decat orice distanta din graf
        self.INF = 1 << 30

        # n = numar de noduri, m = numar de muchii
        self.n = 0
        self.m = 0

        # adj[node] = lista de adiacenta a nodului node
        # perechea (neigh, w) semnifica arc de la node la neigh de cost w
        self.adj: list[list[tuple[int, int]]] = []

        # nodul sursa
        self.source = 0

    def solve(self) -> None:
        self.read_input()
        self.write_output(self.get_result())

    def read_input(self) -> None:
        with open("in", "r") as fin:
            tokens = fin.read().split()
            if not tokens:
                return

            self.n = int(tokens[0])
            self.m = int(tokens[1])
            self.source = int(tokens[2])

            self.adj = [[] for _ in range(self.n + 1)]

            idx = 3
            for _ in range(self.m):
                x = int(tokens[idx])
                y = int(tokens[idx + 1])
                w = int(tokens[idx + 2])
                self.adj[x].append((y, w))
                idx += 3

    def get_result(self) -> list[int]:
        return self.dijkstra(self.source)

    def dijkstra(self, source: int) -> list[int]:
        # Initializam vectorul de distante cu distante infinite.
        d = [self.INF] * (self.n + 1)
        p = [0] * (self.n + 1)

        # Inseram nodul de plecare in coada si ii actualizam distanta.
        d[source] = 0
        pq: list[tuple[int, int]] = [(0, source)]

        while pq:
            # Scoatem top-ul.
            cost, node = heapq.heappop(pq)

            # Daca distanta din coada nu e optima, dam discard.
            if cost > d[node]:
                continue

            # Ii parcurgem toti vecinii.
            for neigh, w in self.adj[node]:
                # Se imbunatateste distanta?
                if d[node] + w < d[neigh]:
                    # Actualizam distanta si parintele.
                    d[neigh] = d[node] + w
                    p[neigh] = node

                    # Inseram noua distanta.
                    heapq.heappush(pq, (d[neigh], neigh))

        # Toate nodurile catre care distanta este inca INF nu pot fi atinse din
        # nodul source, deci le setam pe -1.
        for node in range(1, self.n + 1):
            if d[node] == self.INF:
                d[node] = -1

        return d

    def write_output(self, d: list[int]) -> None:
        with open("out", "w") as fout:
            out_str = " ".join(str(d[node]) for node in range(1, self.n + 1))
            fout.write(out_str + " \n")


def main() -> None:
    task = Task()
    task.solve()


# [ATENTIE] NU modifica functia main!
if __name__ == "__main__":
    main()

File: algorithms/lab08/rust/01-dijkstra/01-dijkstra.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::collections::BinaryHeap;
use std::cmp::Reverse;
use std::fs::File;
use std::io::{BufRead, BufReader, Write};

#[derive(Default)]
struct Task {
    // n = numar de noduri, m = numar de muchii
    n: i32,
    m: i32,

    // adj[node] = lista de adiacenta a nodului node
    // perechea (neigh, w) semnifica arc de la node la neigh de cost w
    adj: Vec<Vec<(i32, i32)>>,

    // nodul sursa
    source: i32,
}

impl Task {
    // numarul maxim de noduri
    const NMAX: usize = 50_005;

    // valoare mai mare decat orice distanta din graf
    const INF: i32 = 1 << 30;

    pub fn solve(&mut self) {
        self.read_input();
        let result = self.get_result();
        self.write_output(&result);
    }

    fn read_input(&mut self) {
        let fin = File::open("in").expect("in");
        let mut lines = BufReader::new(fin).lines().map(|l| l.unwrap());
        let first = lines.next().unwrap();
        let mut tokens = first.split_whitespace().map(|s| s.parse::<i32>().unwrap());
        self.n = tokens.next().unwrap();
        self.m = tokens.next().unwrap();
        self.source = tokens.next().unwrap();
        self.adj = vec![vec![]; Self::NMAX];
        for _ in 0..self.m {
            let line = lines.next().unwrap();
            let mut tok = line.split_whitespace().map(|s| s.parse::<i32>().unwrap());
            let x = tok.next().unwrap();
            let y = tok.next().unwrap();
            let w = tok.next().unwrap();
            self.adj[x as usize].push((y, w));
        }
    }

    fn get_result(&self) -> Vec<i32> {
        return self.dijkstra(self.source);
    }

    fn dijkstra(&self, source: i32) -> Vec<i32> {
        // Initializam vectorul de distante cu distante infinite.
        let mut d = vec![Self::INF; (self.n + 1) as usize];

        // Folosim un BinaryHeap cu Reverse pentru a obtine un min-heap.
        let mut pq = BinaryHeap::new();

        // Inseram nodul de plecare in coada si ii actualizam distanta.
        d[source as usize] = 0;
        pq.push(Reverse((0, source)));

        while let Some(Reverse((cost, node))) = pq.pop() {
            // Daca distanta din coada nu e optima, dam discard.
            if cost > d[node as usize] {
                continue;
            }

            // Ii parcurgem toti vecinii.
            for &(neigh, w) in &self.adj[node as usize] {
                // Se imbunatateste distanta?
                if d[node as usize] + w < d[neigh as usize] {
                    // Actualizam distanta.
                    d[neigh as usize] = d[node as usize] + w;

                    // Inseram noua distanta.
                    pq.push(Reverse((d[neigh as usize], neigh)));
                }
            }
        }

        // Toate nodurile catre care distanta este inca INF nu pot fi atinse din
        // nodul source, deci le setam pe -1.
        for node in 1..=self.n {
            if d[node as usize] == Self::INF {
                d[node as usize] = -1;
            }
        }

        return d;
    }

    fn write_output(&self, d: &[i32]) {
        let mut fout = File::create("out").expect("out");
        for node in 1..=self.n {
            write!(fout, "{} ", d[node as usize]).unwrap();
        }
        writeln!(fout).unwrap();
    }
}

// [ATENTIE] NU modifica functia main!
fn main() {
    let mut task = Box::new(Task::default());
    task.solve();
}
